consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+use blockchain_storage::Transactional;
+use common::chain::block::block_index::BlockIndex;
+use common::chain::block::{calculate_tx_merkle_root, calculate_witness_merkle_root, Block};
+use common::chain::config::ChainConfig;
+use common::chain::{
+    OutPoint, OutPointSourceId, OutputSpentState, SpendablePosition, Spender, Transaction,
+    TxMainChainIndex, TxMainChainPosition,
+};
+use common::chain::{SpendError, TxMainChainIndexError};
+use common::primitives::{time, Amount, BlockHeight, Id, Idable};
+use thiserror::Error;
 mod orphan_blocks;
-// use orphan_blocks::OrphanBlocks;
+use parity_scale_codec::Encode;
+use std::collections::BTreeMap;
 
-// struct Consensus<S: BlockchainStorage> {
-//     chain_config: ChainConfig,
-//     blockchain_storage: S,
-//     orphan_blocks: OrphanBlocks,
-// }
+#[allow(dead_code)]
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum BlockError {
+    #[error("Unknown error")]
+    Unknown,
+    // Orphan block
+    #[error("Orphan")]
+    Orphan,
+    #[error("Invalid block height `{0}`")]
+    InvalidBlockHeight(BlockHeight),
+    #[error("The previous block invalid")]
+    PrevBlockInvalid,
+    #[error("The storage cause failure `{0}`")]
+    StorageFailure(blockchain_storage::Error),
+    #[error("The block not found")]
+    NotFound,
+    #[error("Invalid block source")]
+    InvalidBlockSource,
+    // To be expanded
+}
+
+impl From<blockchain_storage::Error> for BlockError {
+    fn from(_err: blockchain_storage::Error) -> Self {
+        // On storage level called err.recoverable(), if an error is unrecoverable then it calls panic!
+        // We don't need to cause panic here
+        BlockError::Unknown
+    }
+}
+
+impl From<SpendError> for BlockError {
+    fn from(_: SpendError) -> Self {
+        // To be expanded
+        BlockError::Unknown
+    }
+}
+
+impl From<TxMainChainIndexError> for BlockError {
+    fn from(_: TxMainChainIndexError) -> Self {
+        // To be expanded
+        BlockError::Unknown
+    }
+}
+
+// DSA allows us to have blocks up to 1mb
+const MAX_BLOCK_WEIGHT: usize = 1_048_576;
+
+#[allow(dead_code)]
+// TODO: We will generalize it when Lukas will be ready for that. At the moment, he recommended use
+//  types directly.
+// struct Consensus<'a, S: Transactional<'a> + BlockchainStorage> {
+struct Consensus {
+    chain_config: ChainConfig,
+    blockchain_storage: blockchain_storage::Store, //&'a mut S,
+    orphan_blocks: OrphanBlocksPool,
+}
+
+type CachedInputs = BTreeMap<Id<Transaction>, TxMainChainIndex>;
+type PeerId = u32;
+
+#[derive(Copy, Clone, Eq, Debug, PartialEq)]
+pub enum BlockSource {
+    Peer(PeerId),
+    Local,
+}
+
+impl Consensus {
+    fn make_db_tx(&mut self) -> ConsensusRef {
+        let db_tx = self.blockchain_storage.transaction_rw();
+        ConsensusRef {
+            chain_config: &self.chain_config,
+            db_tx,
+            orphan_blocks: &mut self.orphan_blocks,
+        }
+    }
+
+    #[allow(dead_code)]
```

Comment:
Why all the `dead_code` annotations? I suspect most of them could go away if you make the `Consensus` struct public (only the struct itself, not its fields)

Reply:
I think this is just a development relic. Please remove them all, Anton. I think nothing should be left when you're done.

Reply:
Removed! Thank you guys!

Reply:
Removed! Thank you guys!